distributed/worker.py

```diff
@@ -1563,479 +1628,433 @@ def add_task(
+            )
 
-            # clear `who_has` of stale info
-            ts.who_has.clear()
+        if ts.resource_restrictions is not None:
+            for resource, quantity in ts.resource_restrictions.items():
+                self.available_resources[resource] += quantity
 
-        except Exception as e:
-            logger.exception(e)
-            if LOG_PDB:
-                import pdb
+        self.executing_count -= 1
+        self.executed_count += 1
+        recommendations, s_msgs = self._put_key_in_memory(
+            ts, value, stimulus_id=stimulus_id
+        )
+        s_msgs.append(self.get_task_state_for_scheduler(ts))
+        return recommendations, s_msgs
 
-                pdb.set_trace()
-            raise
+    def transition_constrained_executing(self, ts, *, stimulus_id):
+        if self.validate:
+            assert not ts.waiting_for_data
+            assert ts.key not in self.data
+            assert ts.state in READY
+            assert ts.key not in self.ready
+            assert all(
+                dep.key in self.data or dep.key in self.actors
+                for dep in ts.dependencies
+            )
+        for resource, quantity in ts.resource_restrictions.items():
+            self.available_resources[resource] -= quantity
+        ts.state = "executing"
+        self.executing_count += 1
+        self.loop.add_callback(self.execute, ts.key, stimulus_id=stimulus_id)
+        return {}, []
 
-    def transition_fetch_flight(self, ts, worker=None):
-        try:
-            if self.validate:
-                assert ts.state == "fetch"
-                assert ts.dependents
+    def transition_ready_executing(self, ts, *, stimulus_id):
+        if self.validate:
+            assert not ts.waiting_for_data
+            assert ts.key not in self.data
+            assert ts.state in READY
+            assert ts.key not in self.ready
+            assert all(
+                dep.key in self.data or dep.key in self.actors
+                for dep in ts.dependencies
+            )
+        ts.state = "executing"
+        self.executing_count += 1
+        self.loop.add_callback(self.execute, ts.key, stimulus_id=stimulus_id)
+        return {}, []
 
-            ts.coming_from = worker
-            self.in_flight_tasks += 1
-        except Exception as e:
-            logger.exception(e)
-            if LOG_PDB:
-                import pdb
+    def transition_flight_fetch(self, ts, *, stimulus_id):
+        if self.validate:
+            assert ts.state == "flight"
 
-                pdb.set_trace()
-            raise
+        self.in_flight_tasks -= 1
+        ts.coming_from = None
 
-    def transition_flight_fetch(self, ts):
-        try:
-            if self.validate:
-                assert ts.state == "flight"
+        for w in ts.who_has:
+            self.pending_data_per_worker[w].append(ts.key)
+        ts.state = "fetch"
+        heapq.heappush(self.data_needed, (ts.priority, ts.key))
```

Comment:
This is the important change. Every task in state `fetch` is now part of `data_needed`